syntax = "proto3";
package lion_protocol;

enum CommandType {
  GET = 0;
  SET = 1;
}

enum PowerType {
  SHUTDOWN  = 0;
  REBOOT    = 1;
}

enum SysInfoType {
  ARCH          = 0;
  OS            = 1;
  OS_RELEASE    = 2;
  CPU_NUM       = 3;
  CPU_SPEED     = 4;
  STORAGE_TOTAL = 5;
  STORAGE_FREE  = 6;
  UPTIME        = 7;
  TEMPERATURE   = 8;
}

message Power {
  PowerType type = 1;
}

message SysInfo {
  SysInfoType type = 1;
}

message ResourceType {
  oneof resource {
    Power   power   = 1;
    SysInfo sysinfo = 2;
  }
}

message PayloadType {
  oneof payload_val {
    int32        ival   = 1;
    uint32       uval   = 2;
    double       rval   = 3;
    bool         bval   = 4;
    string       sval   = 5;
  }
}

message Request {
  CommandType   command   = 1;
  ResourceType  resource  = 2;
  PayloadType   payload   = 3;
}

enum ResponseStatus {
  OK    = 0;
  FAIL  = 1;
}

message Response {
  CommandType     command   = 1;
  ResourceType    resource  = 2;
  PayloadType     payload   = 3;
  ResponseStatus  status    = 4;
}
